src/cargo/ops/cargo_add/mod.rs

```diff
@@ -537,19 +538,67 @@ fn get_latest_dependency(
                     std::task::Poll::Pending => registry.block_until_ready()?,
                 }
             };
-            let latest = possibilities
-                .iter()
-                .max_by_key(|s| {
-                    // Fallback to a pre-release if no official release is available by sorting them as
-                    // less.
+
+            fn find_latest<'a>(
+                possibilities: impl Iterator<Item = &'a Summary>,
+            ) -> Option<&'a Summary> {
+                possibilities.max_by_key(|s| {
+                    // Fallback to a pre-release if no official release is available by sorting them
+                    // as less.
                     let stable = s.version().pre.is_empty();
                     (stable, s.version())
                 })
-                .ok_or_else(|| {
-                    anyhow::format_err!(
-                        "the crate `{dependency}` could not be found in registry index."
-                    )
-                })?;
+            }
+
+            let mut latest = find_latest(possibilities.iter()).ok_or_else(|| {
+                anyhow::format_err!(
+                    "the crate `{dependency}` could not be found in registry index."
+                )
+            })?;
+
+            if config.cli_unstable().msrv_policy {
+                if let Some(rust_version) = manifest
+                    .data
+                    .as_table()
+                    .get("package")
+                    .and_then(|p| p.get("rust-version"))
+                    .and_then(|v| v.as_str())
+                    .and_then(|v| semver::VersionReq::parse(v).ok())
+                {
+                    // Find the latest version of the dep which has a compatible MSRV. Candidates
+                    // without a rust-version field are treated as compatible.
+                    let latest_msrv = find_latest(possibilities.iter().filter(|s| {
```

Comment:
We already have `latest`, should we check that first before doing more processing?

Alternatively, we could `sort_by` instead of `max_by_key` and then we can just walk the results until we get a compatible version.